source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,201 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+   */
+
+  // Select the sink implementation to use based on type specified in YAML request body
+  switch (type) {
+  case ProtoOutputSinkType::kStreamingAdmin:
+    return std::make_unique<AdminPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kBufferedAdmin:
+    return std::make_unique<BufferedPerTapSinkHandle>(*this);
+  case ProtoOutputSinkType::kFilePerTap:
+  case ProtoOutputSinkType::kStreamingGrpc:
+    PANIC("not implemented");
+  case ProtoOutputSinkType::OUTPUT_SINK_TYPE_NOT_SET:
+  default:
+    PANIC_DUE_TO_CORRUPT_ENUM;
+  }
+}
+
+void AdminHandler::TraceBuffer::bufferTrace(
+    std::shared_ptr<envoy::data::tap::v3::TraceWrapper> trace) {
+  // Ignore traces once the buffer is full or flushed
+  if (flushed() || full()) {
+    return;
+  }
+
+  buffer_->emplace_back(std::move(*trace));
+}
+
+void AdminHandler::AdminPerTapSinkHandle::submitTrace(TraceWrapperPtr&& trace,
+                                                      envoy::config::tap::v3::OutputSink::Format) {
   ENVOY_LOG(debug, "admin submitting buffered trace to main thread");
   // Convert to a shared_ptr, so we can send it to the main thread.
   std::shared_ptr<envoy::data::tap::v3::TraceWrapper> shared_trace{std::move(trace)};
+  // Non owning pointer to the attached request, does not preserve lifetime unless in use
+  std::weak_ptr<AttachedRequest> weak_attached_request(parent_.attached_request_);
+
   // The handle can be destroyed before the cross thread post is complete. Thus, we capture a
   // reference to our parent.
-  parent_.main_thread_dispatcher_.post([&parent = parent_, trace = shared_trace, format]() {
-    if (!parent.attached_request_.has_value()) {
+  parent_.main_thread_dispatcher_.post([weak_attached_request, shared_trace] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      ENVOY_LOG(debug, "attached request does not exist, not streaming trace");
+      return; // No attached request, flushed already
+    }
+
+    attached_request->streamMsg(*shared_trace, false);
+  });
+}
+
+void AdminHandler::BufferedPerTapSinkHandle::submitTrace(
+    TraceWrapperPtr&& trace, envoy::config::tap::v3::OutputSink::Format) {
+  // Convert to a shared_ptr to extend lifetime so we can send it to the main thread.
+  std::shared_ptr<envoy::data::tap::v3::TraceWrapper> shared_trace(std::move(trace));
+  // Non owning pointer to the attached request, does not preserve lifetime unless in use
+  std::weak_ptr<AttachedRequest> weak_attached_request(parent_.attached_request_);
+
+  parent_.main_thread_dispatcher_.post([shared_trace, weak_attached_request] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      ENVOY_LOG(debug, "attached request does not exist, not buffering trace");
```

Comment:
Some duplicate code with above; consider a helper class, maybe an intermediate parent in the class hierarchy?

Reply:
Unless you feel strongly about this I would lean toward leaving this as is. I have a few reasons for this:

1. Most of the duplicate code lies in the lambda function body which is rather short, and all instances of this duplicate code are very close to one another in the code.
2. I can't imagine Envoy will be introducing many more sink types for the admin handler in the future, so the number of instances of duplicate code should stay fairly low
3. I think we may well end up with more code overall which is more confusing to navigate by adding an intermediary class as it would not have an overly well defined "purpose" other than to avoid code duplication
4. We have to be somewhat careful about what is captured by the lambda with respect to object lifetime. In particular, if I added an intermediate class in the SinkHandle hierarchy, we would need to preserve the lifetime of the sink handle to use its functionality in the lambda function. This is not done by default as a new SinkHandle is created per tapped request.

Reply:
Agreed, lgtm.